File: LinkedList/Medium/1669. Merge In Between Linked Lists/1669. Merge In Between Linked Lists.c
#include <stdlib.h>

/**
 * Problem: Merge in Between Linked Lists
 * Description: Given two linked lists list1 and list2, and two integers a and b, merge the two linked lists 
 * by splicing list2 into list1 from position a to b. The nodes between positions a and b in list1 are removed, 
 * and list2 is inserted in their place.
 * 
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */

/**
 * Merges list2 into list1 between positions a and b.
 *
 * @param list1: The head of the first singly-linked list.
 * @param a: The start position (1-based index) in list1 where the merge starts.
 * @param b: The end position (1-based index) in list1 where the merge ends.
 * @param list2: The head of the second singly-linked list to merge into list1.
 * @return: The head of the modified list1.
 */
struct ListNode* mergeInBetween(struct ListNode* list1, int a, int b, struct ListNode* list2) {
    struct ListNode *temp1 = list1, *temp2 = list2, *leftPrev = NULL, *prev = NULL;

    // Traverse list1 to find the node just before position 'a'
    for (int i = 0; i < b; i++) {
        if (i == a - 1) {
            leftPrev = temp1;
        }
        temp1 = temp1->next;
    }

    // Link the node before 'a' to the head of list2
    leftPrev->next = list2;

    // Traverse to the end of list2
    while (temp2 != NULL) {
        prev = temp2;
        temp2 = temp2->next;
    }

    // Connect the end of list2 to the node after position 'b' in list1
    prev->next = temp1->next;

    return list1;
}

## Time Complexity: O(N + M), where N is the number of nodes in list1, and M is the number of nodes in list2.
## Space Complexity: O(1), as we are using only a constant amount of extra space.

File: LinkedList/Medium/1669. Merge In Between Linked Lists/1669. Merge In Between Linked Lists.py
# Problem: Merge Two Linked Lists with an Insertion Point
# Description: Given two singly-linked lists `list1` and `list2`, and two integers `a` and `b`, 
# merge `list2` into `list1` starting at position `a` and ending at position `b`. 
# The nodes between positions `a` and `b` in `list1` should be replaced by `list2`.

# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution(object):
    def mergeInBetween(self, list1, a, b, list2):
        """
        Merges list2 into list1 starting at position `a` and ending at position `b`.

        :param list1: ListNode, the head of the first singly-linked list.
        :param a: Integer, the start position (0-indexed) for merging list2 into list1.
        :param b: Integer, the end position (0-indexed) for merging list2 into list1.
        :param list2: ListNode, the head of the second singly-linked list to be merged into list1.
        :return: ListNode, the head of the modified linked list.
        """
        # Create a dummy node to simplify edge cases
        list1 = ListNode(-1, list1)
        temp = list1
        
        # Traverse to the node just before position `a`
        for i in range(a):
            temp = temp.next
        leftNode = temp
        
        # Traverse to the node just after position `b`
        for j in range(b - a + 1):
            temp = temp.next
        
        # Find the last node of list2
        temp_2 = list2
        prev = None
        while temp_2:
            prev = temp_2
            temp_2 = temp_2.next
        
        # Merge list2 into list1
        prev.next = temp.next
        leftNode.next = list2
        
        return list1.next

## Time Complexity: O(N + M), where N is the number of nodes in list1 and M is the number of nodes in list2.
## Space Complexity: O(1), as we only use a constant amount of extra space.

# Example usage:
# list1 = ListNode(0)
# list1.next = ListNode(1)
# list1.next.next = ListNode(2)
# list1.next.next.next = ListNode(3)
# list1.next.next.next.next = ListNode(4)
# list2 = ListNode(1000000)
# list2.next = ListNode(1000001)
# solution = Solution()
# result = solution.mergeInBetween(list1, 2, 4, list2)
# print(result)  # Expected output: 0 -> 1 -> 1000000 -> 1000001
